Make first rc

JelF opened the pull request:
Implemented, styled and documented Verifier (except flow documentation and README.md)

JelF committed:
refactored logic

JelF committed:
perfect rubocop and code style

JelF committed:
added docs

JelF committed:
added travis

akxcv commented:
I think you should configure Travis to only build the pull request once (e.x. only build pushes to "master" branch and PRs).

JelF commented:
@akxcv travis made exactly one build when i pushed this branch

akxcv commented:
@JelF and it made another one when you created this pull request.

JelF committed:
class_builder_spec and rubocop-rspec features

JelF committed:
added applicator_spec, renamed binding to binding_

JelF committed:
applciator_with_otions_spec

JelF committed:
styles in verifier_spec

JelF committed:
added XVerifier::VERSION spec

JelF committed:
updated rubocop-rspec, updated styling

JelF commented:
@akxcv done

JelF commented:
ill prefer to finalize work on this point and then add changes via separate PR's @akxcv, could you review entire repository in this PR?

JelF committed:
updated travis config

- use with newest bundler
- use ruby 2.3.4 for 2.3 family
- add ruby 2.4.1 for 2.4 family

akxcv commented on .rubocop.yml:
```diff
@@ -14,3 +17,15 @@ Metrics/BlockLength:
 Lint/AmbiguousBlockAssociation:
   # REVIEW: why is that dangerous?
```
Not sure about the dangers, but I think that `some_method(a) { |val| puts val }` is more readable than `some_method a { |val| puts val }`.

JelF replied:
It is about following syntax:
```
spec/xverifier/verifier_spec.rb:55:9: W: Parenthesize the param lambda to make sure that the block will be associated with the verify method call.
        verifier.verify ->(context) { message!(context) }
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

akxcv replied:
It wants you to put parens around the lambda...
Yeah, I guess we can do without this rule.

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
 
 module XVerifier
   # @abstract implement `#call`
+  # Applicates objects called `applicable`
```
"Applies "applicable" objects to given bindings" or smth like that

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
 module XVerifier
   # @abstract implement `#call`
+  # Applicates objects called `applicable`
+  # (all objects are applicable, this typing made for usage purposes).
```
"this naming(?) *is chosen* for usage(convenience?) purposes"

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
+  # Applicates objects called `applicable`
+  # (all objects are applicable, this typing made for usage purposes).
+  #
+  # This class uses ClassBuilder subsystem, start with .call method
```
Maybe just "system", not "subsystem"?
Also, "When reading the code, we suggest starting from '.call' method"

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
-        ProcApplicatior
-      else
-        Quoter
+    # Proxy is used when applicable itself is an Applicator.
```
"when applicable is itself an instance of Applicator" sounds clearer

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
-      else
-        Quoter
+    # Proxy is used when applicable itself is an Applicator.
+    # It just delegates #call to it
```
"It simply delegates "#call" method to the applicable."

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
+    #   # => Applicator.call(:foo, binding_, context)
+    class Proxy < self
+      # @param applicable [Applicator]
+      # @return Proxy if applicable is Applicator
```
"is an instance of Applicator"

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
 
-    def initialize(applicable)
-      self.applicable = applicable
+      # @param binding_ [#instance_exec] target to applicate applicable
```
"target to apply the applicable to"
It's beginning to sound weird, but that's just a consequence of naming.

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
-    def initialize(applicable)
-      self.applicable = applicable
+      # @param binding_ [#instance_exec] target to applicate applicable
+      # @param context additional info to send it to applicable
```
"to send to the applicable", you don't need the "it" here

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
-    #   @param context
-
-    private
+    # MethodExtractor is used when applicable is a symbol.
```
"the applicable"
Also, "Symbol"

JelF committed:
fixed bug with MethodExtractor on Binding binding

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
-
-    private
+    # MethodExtractor is used when applicable is a symbol.
+    # It extracts extracts method from binding_ and executes it on binding_
```
"It extracts a method from binding_ and then executes it in its context" or smth

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
-        binding.instance_exec(&lambda)
-      else
-        binding.instance_exec(context, &lambda)
+      # @param binding_ [#instance_exec] target to applicate applicable
```
Same as earlier

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
         end
       end
 
+      # @return [String, Integer] file and line where `Applicator.call` called
```
"was called", or "is called", or "is being called"... Depending on what you actually meant

JelF replied:
i also forgot to make a spec for this

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
       end
     end
 
+    # Quoter is used when there is no other way to applicate applicatable.
```
Here and everywhere else, change "applicate" to "apply".

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
+
+    attr_accessor :applicable
+
+    # Applicates applicable on binding_ with context
```
And "applicates" to "applies".

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
+    # Applicates applicable on binding_ with context
+    # @todo add @see #initialize when it's todo done
+    # @param applicable [applicable]
+    #   see examples in sublcasses defenitions
```
"in subclass definitions"
Or, "in definitions of subclasses"

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
+    #   where it would be `instance_exec`uted, or a binding_
+    # @param context
+    #   geneneric data you want to pass to applicable function.
+    #   If applicable would not accept params, it would not be sent
```
"If applicable cannot accept params, the data will not be sent"

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
+    # @!method call(binding_, context)
+    #   @abstract
+    #   Applicates applicable on binding_ with context
+    #   @param binding_ [#instance_exec] biding would be used in application
```
"binding to be used for application"

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
+    #   @abstract
+    #   Applicates applicable on binding_ with context
+    #   @param binding_ [#instance_exec] biding would be used in application
+    #   @param context param would be passed if requested
```
"param that will be passed if requested"

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
+
+    private
+
+    # invokes lambda basing on it's arity
```
"based on its arity"

JelF replied:
Doesn't it mean like "select some lambda"
Used: "invokes lambda respecting it's arity"

akxcv commented on lib/xverifier/applicator.rb:
```diff
@@ -2,77 +2,181 @@
+    # @param [Proc] lambda
+    # @param binding_ [#instance_exec] binding_ would be used in application
+    # @param context param would be passed if lambda arity > 0
+    # @return invokation result
```
"invocation"

akxcv commented on lib/xverifier/applicator_with_options.rb:
```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module XVerifier
+  # An applicator with usefull options
```
useful

akxcv commented on lib/xverifier/applicator_with_options.rb:
```diff
@@ -0,0 +1,43 @@
+  # @attr action [Applicator]
+  #   main action to applicate on call
+  # @attr if_condition [Applicator]
+  #   main action only applicate if this applicates to truthy value
```
"...only applies if this evaluates..."

JelF replied:
evaluates is more generic and does not describe, that this stuff would be applicated using the rules of application

akxcv commented on lib/xverifier/applicator_with_options.rb:
```diff
@@ -0,0 +1,43 @@
+    # @param context
+    #   generic context to applicate (see Applicator)
+    # @return main action application result
+    # @return [nil] if conditions check failed
```
condition checks

akxcv commented on lib/xverifier/class_builder.rb:
```diff
@@ -0,0 +1,55 @@
+
+module XVerifier
+  # ClassBuilder is something like Uber::Builder, but it uses
+  # inverted dependencies. It means, that child classes are
```
dependency injection / dependency inversion?

akxcv commented on lib/xverifier/class_builder.rb:
```diff
@@ -0,0 +1,55 @@
+module XVerifier
+  # ClassBuilder is something like Uber::Builder, but it uses
+  # inverted dependencies. It means, that child classes are
+  # deciding to be or not to be used
```
"It means that child classes decide whether they should be built or not"

akxcv commented on lib/xverifier/class_builder.rb:
```diff
@@ -0,0 +1,55 @@
+  # inverted dependencies. It means, that child classes are
+  # deciding to be or not to be used
+  # @attr klasses [Array(Class)]
+  #   klasses to iterate during search of most suitable
```
classes

akxcv commented on lib/xverifier/class_builder.rb:
```diff
@@ -0,0 +1,55 @@
+    # Mixin provides usefull methods to integrate builder subsystem.
+    # Feel free to override or just never include it.
+    # @attr_writer [Array(Class)] buildable_classes
+    #   Array of classes which would be checked if they
```
will

akxcv commented on lib/xverifier/class_builder.rb:
```diff
@@ -0,0 +1,55 @@
+    # Feel free to override or just never include it.
+    # @attr_writer [Array(Class)] buildable_classes
+    #   Array of classes which would be checked if they
+    #   suite constructor arguments. Order maters
```
matters

akxcv commented on lib/xverifier/class_builder.rb:
```diff
@@ -0,0 +1,55 @@
+        @class_builder = ClassBuilder.new(klasses).freeze
+      end
+
+      # Default implementation if build_class. Feel free to change
```
of

akxcv commented on lib/xverifier/verifier.rb:
```diff
@@ -10,12 +10,17 @@ module XVerifier
   # @attr messages [Array]
   #   Array to collect all messages yielded by verifier
```
Array with all messages yielded by the verifier

akxcv commented on lib/xverifier/verifier/applicator_with_options_builder.rb:
```diff
@@ -0,0 +1,74 @@
+
+module XVerifier
+  class Verifier
+    # Builds ApplicatorWithOptions from different invokation styles.
```
invocation

akxcv commented on lib/xverifier/verifier/applicator_with_options_builder.rb:
```diff
@@ -0,0 +1,74 @@
+    # @attr base [Class]
+    #   class for which applicator_with_options should be built
+    # @attr args [Array]
+    #   array of arguments invoked with
```
an array of arguments to invoke methods with ?

JelF replied:
it changes sence, but should be okish

JelF replied:
well, not in this place. I'll try to reword:
array of arguments `Verifier.verify` invoked with

akxcv commented on lib/xverifier/verifier/applicator_with_options_builder.rb:
```diff
@@ -0,0 +1,74 @@
+    # @attr args [Array]
+    #   array of arguments invoked with
+    # @attr block [Proc]
+    #   block invoked with
```
-||-

JelF replied:
???

akxcv replied:
I meant "same as above".
"a block to invoke methods with"

akxcv commented on lib/xverifier/verifier/applicator_with_options_builder.rb:
```diff
@@ -0,0 +1,74 @@
+        new(base, args, block).call
+      end
+
+      # Tries different invokation styles untill one matches
```
invocation; until

akxcv commented on lib/xverifier/verifier/applicator_with_options_builder.rb:
```diff
@@ -0,0 +1,74 @@
+        verifier, *rest = args
+        return unless verifier.is_a?(Class)
+        raise ArgumentError, <<~ERROR unless verifier < base
+          Nested verifiers should inherit verifier they nested in
```
"...should inherit from the verifier they are nested in"

akxcv commented on spec/xverifier/applicator_spec.rb:
```diff
@@ -0,0 +1,120 @@
+  subject(:applicator) { described_class.build(applicable) }
+
+  let(:binding_class) do
+    Class.new { define_method(:foo) { |*| } }
```
What on Earth is this magic?

akxcv commented on spec/xverifier/class_builder_spec.rb:
```diff
@@ -0,0 +1,25 @@
+    Class.new { extend XVerifier::ClassBuilder::Mixin }
+  end
+
+  let(:unbuildabe) { class_double(klass, build_class: nil) }
```
unbuildable

akxcv commented on spec/xverifier_spec.rb:
```diff
@@ -0,0 +1,14 @@
+      expect(XVerifier::VERSION).to match(/\A\d+\.\d+\.\d+(_.+)?\z/)
+    end
+  end
+end
```
❤️

akxcv reviewed (changes requested):
Overall looks good! 👍

JelF committed:
review fixes

JelF committed:
specs for InstanceEvaluator#caller_line

JelF commented:
@akxcv can you re-check it?

JelF committed:
Added readme

akxcv commented on .rubocop.yml:
```diff
@@ -14,3 +17,15 @@ Metrics/BlockLength:
+    XVerifier: xverifier
+
+RSpec/MessageSpies:
+  # Spies are not always usable, i.e. we wanna to stub something we haven't create
```
1) "wanna" = "want to"
2) "... haven't created"

akxcv reviewed (commented):
Alright, let's correct this one line and then I'll make a pull request with grammar corrections for the README (eventually, if I find any mistakes).

JelF committed:
grammar
